[Wee 7] Security login

sunggyupaik opened the pull request:
Spring Security 구현하기

sunggyupaik committed:
스프링 보안을 이용하여 인증을 하도록  수정하라

인증이 필요한 요청에 대해서 Interceptor로 요청에 대해 식별하하도록
되어있었습니다. 하지만, 단순한 처리는 웹 보안이 취약하며, 비밀번호 또한
암호화하지 않아 노출 될 위험이 있습니다. 따라서, 스프링 보안을 이용하여
비밀번호를 암호화하며 인증을 하도록 수정하였습니다.

sunggyupaik committed:
보안을 제외하고 이전 과제와 합쳐라

현재 새로운 7주차 과제의 내용만 이용하여 보안이 구현되어
있었습니다. 6주차에 있던 내용과 흐름을 잇는 것이 더 좋다고 생각합니다.
따라서, 보안을 제외한 모든 ㅜ분을 6주차 과제로 합칩니다. 테스트 부분은
아직 합치지 않습니다.

sunggyupaik committed:
사용자 생성 이후 반환되는 정보를 수정하라

사용자를 생성하면 모든 사용자의 정보가 리턴하도록 되어있었습니다.
하지만, 비밀번호까지 리턴하여 보여줄 필요가 없으며, 암호화하여 저장하기
때문에 null을 리턴합니다. 따라서, 아이디, 이름, 이메일만 리턴하도록
수정하였습니다.

sunggyupaik committed:
사용자 테스트를 수정하라

사용자 테스트는 비밀번호를 암호화하지 않은 상태로 되어있었습니다.
하지만, 이제 사용자는 비밀번호를 암호화하여 저장하도록 하였습니다.
따라서, 비밀번호에 대한 암호화에 대해 확인할 수 있도록 추가하였습니다.

sunggyupaik committed:
사용자 테스트에서 비밀번호 변경을 수정하라

사용자의 이전 비밀번호에 대한 암호화를 따로 하고 같은지 비교하고
있었습니다. 하지만, 암호화 하면 같은 비밀번호도 암호화 결과가 다를 수
있어서 올바른 답이 나오지 않습니다. 따라서, 비밀번호가 이전과 같지
않다는 방식으로 검증을 수정했습니다.

sunggyupaik committed:
사용자 인증 테스트에서 비밀번호를 암호화하라

비밀번호가 암호화되지 않고 기존의 입력 그대로 테스트되고 있었습니다.
하지만 테스트에서 암호화 알고리즘을 사용하지 않았다는 오류가 발생합니다.
따라서 비밀번호를 암호화해서 사용하여 테스트를 통과했습니다.

sunggyupaik committed:
사용자 수행에 대한 테스트를 수정하라

사용자의 수행을 테스트합니다. 비밀번호가 암호화하지 않은 상태이므로
암호화하여 진행하도록 진행합니다.

sunggyupaik commented:
일주일 간 잘 부탁드립니다 ㅎㅎ

이전 과제와 현재 과제를 합치는 과정에서 로직을 완성했는데, 테스트를 추가적으로 작성중입니다

wwek7 에서 기본적으로 들어있던 테스트를 모두 주석처리하니 굉장히 지저분한 상태입니다... 가독성에 대해 죄송합니다

sunggyupaik committed:
삭제된 사용자가 주어진 테스트를 수정하라

사용자는 삭제된 식별자에 대해 검사하지 않고 있었습니다.
하지만, 식별자에 대해 사용자가 존재하여도 삭제되면 사용할 수 없습니다.
따라서. 삭제된 사용자를 고려하여 테스트를 추가합니다.

sunggyupaik committed:
인증에 대한 테스트를 수정하라

인증에 대한 테스트는 비밀번호가 암호와되어 있지 않았습니다.
하지만, 비밀번호를 암호화하기 때문에 테스트가 깨지는 문제가 있었습니다.
따라서, 테스트에 암호와에 대한 내용을 테스트할 수 있도록 수정했습니다.

sunggyupaik committed:
로그인 테스트 설명을 수정하라

로그인 테스트에서 토큰 생성의 메서드 설명이 login으로 되어있었습니다.
하지만, 이미 리팩토링 과정에서  createToken으로 변경했었습니다.
따라서, login에 대한 변수와 설명문을 createToken으로 수정합니다.

sunggyupaik committed:
토큰 해석의 반환 정보를 수정하라

토큰 정보는 응답 전용 Data에 담겨져서 반환되고 있었습니다.
하지만, 해당 반환형은 테스트 인증을 어렵게 합니다. 또한 의미가 분명하지
않아 직관적인 이해가 힘듭니다. 따라서, 문자열로 바로 반환하게
하였습니다.

sunggyupaik committed:
상품 요청에 대한 테스트를 수정하라

상품 요청은 필터에 대한 검사를 진행하지 않도록 되어 있었습니다.
필터에서 토큰 해석의 반환 형태가 변경됨에 따라 테스트도 변해야 합니다.
따라서, 상품 요청에 필요한 토큰 해석의 방법을 수정하였습니다.

sunggyupaik committed:
사용자 생성 테스트를 수정하라

사용자 생성 테스트의 예상 결과가 id값이 없도록 되어 있었습니다.
하지만, 사용자는 새로 만들어지면 id가 생기므로 이 id가 정상적으로
생성되었는지 확인해야합니다. 따라서, 예상 결과가 id를 올바로 가질 수
있도록 수정하였습니다.

sunggyupaik committed:
사용하지 않는 파일을 삭제하라

루트 폴더에 사용하지 않는 파일이 같이 저장되어 잇었습니다.
따라서, 해당 파일을 삭제하였습니다.

sunggyupaik committed:
사용자 인증 리턴 방식을 수정하라

현재, 인증된 사용자를 리턴할 때 ofAll의 메서드를 사용하도록
되어있었습니다. 하지만, Effective Java 3판의 Item 1에서는 하나의
매개변수가 주어지면 of를 사용하도록 권고하고 있습니다. 따라서 of를
사용하도록 수젛압니다.

sunggyupaik committed:
토큰 생성 리턴 메서드명을 수정하라

이전에 from을 사용하여 정적 팩터리 메서드로 생성된 토큰을 리턴하도록
되어있었습니다. 하지만, Effective Java 3판에서는 1개의 매개변수에는 of를
사용하도록 권고하고 있습니다. 따라서 from을 of로 수정합니다.

sunggyupaik committed:
토큰 해석 주석문을 수정하라

토큰 해석 시, 서명이 잘못된 경우 예외가 발생한다고 되어있었습니다.
하지만 결국 발생하는 예외는 유효하지 않은 토큰에 대한 예외이므로, 예외에
따라 내용을 적는 것이 일관성이 있다고 판단하였습니다. 따라서, 예외
설명문을 수정합니다. 또한 반환 값은 이메일이라고 되어 있었습니다.
하지만, 이는 너무 구체적인 정보라고 생각되어집니다. 따라서 사용자
정보라른 표현으로 수정하였습니다.

sunggyupaik committed:
컨트롤러와 서비스에 전체 주석을 추가하라

대부분의 컨트롤러와 서비스는 각각 메서드에는 주석이 있었지만 클래스
설명에 대한 주석은 없었습니다. 각 클래스가 어떤 역할을 하는지 확인하는
것도 도움이 된다고 생각합니다. 따라서, 각 컨트롤러와 서비스클래스의
상단에 어떤 역할을 하는지 추가하였습니다.

sunggyupaik committed:
사용자 요청에 대한 주석을 수정하라

예를 들어, 사용자 요청에 대한 리턴을 "...삭제하고 삭제된 사용자"라는
표현을 사용하도록 되어있었습니다. 같은 단어를 중복하는 것보다 다른
단어를 사용하는것이 매끄럽다고 생각합니다. 따라서 "...삭제하고 해당
사용자"라는 단어로 일괄 수정하였습니다. 또한 조사를 빠뜨린 설명문에
조사를 추가하였습니다.

sunggyupaik committed:
주석의 양식의 수정하라

이전에, 생성, 수정, 삭제에 대한 주석 양식이 각기 다르게 되어있었습니다.
따라서, 비슷한 행동에도 각각 다른 주석들을 가지고 있어서 가독성이
떨어졌습니다. 따라서, 이해가 쉽도록 주석의 양식을 통일하였습니다.

sunggyupaik commented:
이번과제를 하면서 여러가지를 생각해보았습니다.~!!

현재 아래와 같은 조회문을
![image](https://user-images.githubusercontent.com/45138206/110777525-8dc9eb80-82a4-11eb-8def-30785353c3bf.png)

아래와 같이 바꾼다면
![image](https://user-images.githubusercontent.com/45138206/110777683-bb169980-82a4-11eb-80f6-1186a8986169.png)

가독성이 어떠신가요?

sunggyupaik commented:
- 필터에 대한 테스트를 진행할 수 있을까요? 필터"만을" 위한 테스트는 없고 Controller에서 확인해야 할까요?

-  `Dozer`는 좋은 라이브러리이지만 `Dto` <-> `Entity` 변환 시, 정적 팩터리 메서드와 사용하니 같은 목적 아래에서 구현방법이 중복되지 않나 생각이 듭니다...  굳이 한쪽만 사용하려고 하는데, 이쯤에서 `Dozer`를 놓아줘도 될까요?
![image](https://user-images.githubusercontent.com/45138206/110781082-ccfa3b80-82a8-11eb-8bea-bfe4d3aced63.png)

- 어디까지 `Dto`로 만들어주어야 할까요?
![image](https://user-images.githubusercontent.com/45138206/110785827-bbb42d80-82ae-11eb-8c32-19348011fa0a.png)
`parseToken`의 메서드의 경우 오른쪽과 같이 정의하면, 제 실력의 한계인지, 기존에 되었던 테스트들이 꺠지게 되어 왼쪽으로 고쳐서 해결했습니다. (변환이 제대로 되지 않아 null이 발생합니다.) 
`Service`에 정의된 메서드들이 Dto를 반환하는 것이 좋지 않나 생각합니다.

sunggyupaik commented:
![image](https://user-images.githubusercontent.com/45138206/110787404-9d4f3180-82b0-11eb-9b9b-55f6251e9cf7.png)

현재 `csrf`와 `frameOptions` 모두 `disable()`로 되어있네요, 이렇게 쓰는것은 다소 아쉬운것 같습니다. 더 찾아보고 적용해볼 수 있을 것 같습니다.


또한 갈수록 코드가 늘어나고, 변수의 양과 가로 길이가 길어지고 있는데요, 이번에 배송 온 클린코드 5장의 `형식 맞추기` 와 더불어 가이드라인을 통해서도 좀 더 가독성 좋은 모양으로 개선해보겠습니다.

johngrib commented:
> 이번과제를 하면서 여러가지를 생각해보았습니다.~!!
> 
> 현재 아래와 같은 조회문을
> ![image](https://user-images.githubusercontent.com/45138206/110777525-8dc9eb80-82a4-11eb-8def-30785353c3bf.png)
> 
> 아래와 같이 바꾼다면
> ![image](https://user-images.githubusercontent.com/45138206/110777683-bb169980-82a4-11eb-80f6-1186a8986169.png)
> 
> 가독성이 어떠신가요?

제가 보기에는 둘 다 큰 차이 없습니다. 저는 후자가 취향이군요.

sunggyupaik commented:
![image](https://user-images.githubusercontent.com/45138206/110789400-2ebfa300-82b3-11eb-9053-813ac1d324d3.png)
```java
.andExpect(content().string(containsString("Product and request")))
```
이 부분을 주석처리한 이유는 given에서 정의한 `ProductBadRequestException`의 예외 메세지를 리턴하지 않기 때문입니다.

아래 사진에서 응답결과를 보시면 빨간색 박스에 `name 값은 필수입니다`라고 적혀있습니다. 제가 정의한 메세지는 `Product not found`였는데, 다른 문자열을 리턴하고 있습니다.
![image](https://user-images.githubusercontent.com/45138206/110789662-78a88900-82b3-11eb-9f6d-1b9390861770.png)

알고보니 Controller에서 `@Valid`에 대한 검사가 진행되어 `@NotBlank`의 message를 표시하기 때문이었습니다.
![image](https://user-images.githubusercontent.com/45138206/110789863-b0173580-82b3-11eb-8844-5a471a234383.png)

그렇다면 제가 정의한 given의 사용은 전혀 연관성이 없는건데, given을 삭제하고 테스트해도 될까요?

sunggyupaik commented:
이번 과제에서 무엇을 좀 더 얻어갈 수 있을까요? 
![image](https://user-images.githubusercontent.com/45138206/110790709-b954d200-82b4-11eb-8c16-a2a3d0f746c6.png)

토큰 만드는 과정에서 `subject`만 사용한 것도 아쉬울 수 있네요, 다른 부분을 적용하는 것도 찾아보면 좋을 것 같습니다

sunggyupaik commented:
ProductController에 `@PreAuthorize("isAuthenticated() and hasAuthority('USER')")` 이런 어노테이션이 있는 것들은 어떻게 테스트를 할 수 있을까요? User 권한에 대한 것을 어떻게 표현하면 좋을지 모르겠습니다.

johngrib commented:
> * 필터에 대한 테스트를 진행할 수 있을까요? 필터"만을" 위한 테스트는 없고 Controller에서 확인해야 할까요?

입출력에 대한 통합 테스트로 굴려버리는 것이 편하겠죠.

johngrib commented:
> * `Dozer`는 좋은 라이브러리이지만 `Dto` <-> `Entity` 변환 시, 정적 팩터리 메서드와 사용하니 같은 목적 아래에서 구현방법이 중복되지 않나 생각이 듭니다...  굳이 한쪽만 사용하려고 하는데, 이쯤에서 `Dozer`를 놓아줘도 될까요?

저는 Dozer를 안 좋아해서 안 씁니다. Dozer 없이 작업하는 쪽이 좋을 것 같으시다면 해보세요.

johngrib commented:
> * 어디까지 `Dto`로 만들어주어야 할까요?
>   ![image](https://user-images.githubusercontent.com/45138206/110785827-bbb42d80-82ae-11eb-8c32-19348011fa0a.png)
>   `parseToken`의 메서드의 경우 오른쪽과 같이 정의하면, 제 실력의 한계인지, 기존에 되었던 테스트들이 꺠지게 되어 왼쪽으로 고쳐서 해결했습니다. (변환이 제대로 되지 않아 null이 발생합니다.)
>   `Service`에 정의된 메서드들이 Dto를 반환하는 것이 좋지 않나 생각합니다.

Dto라고 생각하지 말고 필요한 개념을 전달한다고 생각하고 클래스를 정의해 작업해 보세요. Dto라는 단어를 너무 인식하지 마세요.

johngrib commented:
> 그렇다면면 저 given의 사용은 전혀 연관성이 없는건데, 그렇다면 given을 삭제하고 그냥 사용해도 괜찮을까요?

필요없는 건 지워야죠.

johngrib commented:
> 이번 과제에서 무엇을 좀 더 얻어갈 수 있을까요?

과제는 보안이지만 제가 7주차에 주안점을 두는 것은 커밋 메시지입니다. 과제에서는 security 에 대해 실습을 해보는 거고, PR리뷰에서 커밋 메시지와 커밋 단위에 대해 이야기해봅시다.

johngrib commented:
> ProductController에 `@PreAuthorize("isAuthenticated() and hasAuthority('USER')")` 이런 어노테이션이 있는 것들은 어떻게 테스트를 할 수 있을까요? User 권한에 대한 것을 어떻게 표현하면 좋을지 모르겠습니다.

이럴 땐 스프링 시큐리티 공식 문서를 찾아 봅시다. 직접 찾아보시면 더 좋았겠지만(여기에 링크를 남길지 말 지 고민이 컸습니다!) 벌써 목요일이고 과제에 더 집중하실 필요도 있을 것 같아 링크를 남깁니다.

https://docs.spring.io/spring-security/site/docs/4.0.x/reference/html/test-mockmvc.html#test-mockmvc-securitycontextholder-rpp
![image](https://user-images.githubusercontent.com/1855714/110794269-e99e6f80-82b8-11eb-8fec-4e11bae8f1d9.png)

johngrib reviewed (commented):
e0b8be892608605e37887bdb83cb015897f5dae7 커밋 메시지가 좋습니다. 커밋 메시지 본문도 훌륭합니다. 좀 더 심플하면 더 좋았겠지만 지금 상태로도 좋습니다.
![image](https://user-images.githubusercontent.com/1855714/110794380-0a66c500-82b9-11eb-8961-605e8867aebf.png)

johngrib reviewed (commented):
ef1afc4247d3856b9a78dc4ba509c932049f44dd 커밋 메시지가 좋습니다.

변경 전 상태에 대한 설명, 문제 상황에 대한 설명, 변경한 내용에 대한 설명

3박자가 잘 맞습니다.
![image](https://user-images.githubusercontent.com/1855714/110794829-7fd29580-82b9-11eb-9142-59d9f00236ea.png)

sunggyupaik committed:
요청에 대한 테스트 구조를 수정하라

상품, 사용자, 로그인 요청에 대한 요청 테스트는 옆으로 행을 늘어뜨리는
구조로 되어 있었습니다. 하지만, 이는 행이 100자가 넘어가는 결과를
초래합니다. 따라서 100자 이하로 줄이기 위해 적절하게 라인을 나누도록
수정했습니다.

sunggyupaik committed:
상품 요청 테스트의 불필요한 부분을 삭제하라

기존에 잘못된 상품 요청에 대한 예외를 given에 선언하도록 되어 있었습니다.
하지만, 실제 테스트에서는 해당 예외를 던지지 않고 다른 예외를 던집니다.
따라서, 필요하지 않은 선언이라고 생각되어 삭제합니다.

sunggyupaik committed:
Dozer를 제거하라

기존에 Dozer를 사용하여 사용자와 상품의 생성, 수정, 삭제에 활용했었습니다.
하지만, Dozer와 방법만 다를 뿐 일반 메서드를 통한 처리가 공존했었습니다.
다른 사람들이 보기에 이해가 쉽도록 일반 메서드를 사용하기로 했습니다.
따라서, Dozer의 내용을 모두 제거하고 build.gradle에만 남겨두었습니다.

sunggyupaik commented:
> Dto라고 생각하지 말고 필요한 개념을 전달한다고 생각하고 클래스를 정의해 작업해 보세요.

이 부분에 대해서는 무조건 Dto로 보내야돼! 라는 의미보다 최대한 Entity나 주요 정보를 그대로 보내지 말고 request, response로 변환해서 보내자! 라는 생각으로 했었습니다. 단, 꼭 정해직 규칙이 있는 것이 아니기에, 너무 집착하지 않도록 하겠습니다.

sunggyupaik commented:
커밋에 작성에 대해 쭉 다시 써보면서, 코멘트 달아주신 2개의 메세지를 다시 돌아보았습니다. 그리고 개선점도 찾아보았습니다.

### 1. 제목
 - **종결 동사**

 [좋은 git commit 메시지를 위한 영어 사전](https://blog.ull.im/engineering/2019/03/10/logs-on-git.html?fbclid=IwAR110Dikegni7hOOHOG2nXZD1SGuJWruYOmLsdzPFC1P1yY9ivC0JFYQuC4) 에 보면, `FIX`, `ADD`, `REMOVE`, `REFACTOR`, `CORRECT`, `RENAME`, `VERIFY` 등등 다양한 명령형 동사들에 대해 소개하고있습니다. 

저는 이번 과제에서 주로 `...수정하라`를 많이 사용했는데, `고쳐라` `제거하라` `리팩토링하라` `교정하라` `변경하라` 등등 응용하면 좀 더 구분이 가능할 것 같습니다.

또한 제목 앞에 간단하게 설명을 넣는 것도 생각할 수 있습니다.
`수정: 사용자 테스트를 수정하라`
`추가: 사용자 인증 내용을 추가하라`

e0b8be8
`스프링 보안을 이용하여 인증을 하도록 수정하라`
=> `스프링 보안을 이용하여 인증을 하도록 변경하라`
**interceptor를 spring security방식으로 바꿨기 때문에 변경이 더 어울리지 않나 생각합니다.**

 - **추상화 정도**

제목만 보고도 어떤 변경이 있었는지 파악이 되면 좋겠습니다.

ef1afc4
`사용자 테스트를 수정하라`
=> `비밀번호를 암호화하여 테스트 하도록 수정하라`
**사용자 테스트 중에 "무엇을" 수정하려고 하는지 좀 더 파악이 될 것 같습니다.**
경우에 따라 보완하라, 개선하라 등의 표현도 괜찮아 보입니다.

혹은 취향에 따라 
`비밀번호를 암호화하여 테스트 하도록 수정` 
의 종결도 나쁘지 않습니다.

### 2. 내용
- **변경 전, 변경 과정, 변경 결과**

이전 상태를 충분히 드러내도록 설명을 합니다.
`이전에`, `기존에는`, `A이었습니다`, `B로 되어 있었습니다` 

이후 변경을 왜 하게 되었는지에 대해서 이유를 설명합니다.
`하지만,` `그럴경우`, `A의 문제가 있습니다.` `B가 C로 변경되었습니다.` `D에 취약합니다` `E의 위험이 있습니다` 

마지막에는 수정 결과를 자연스럽게 전달하며 마무리합니다.
`따라서`, `그렇기 때문에`, `A를 B로 수정합니다`, `C를 제거합니다`, `D에 E를 보완합니다` 

혹여나 또다시 변경될 여지가 있다면 마지막에 [한계점](https://wiki.openstack.org/wiki/GitCommitMessages#Information_in_commit_messages)을 남기는 것도 좋을 것 같습니다. 
![image](https://user-images.githubusercontent.com/45138206/110910068-3420fa00-8354-11eb-9015-8ccf0f2e83b0.png)

`A는 B에 따라서 변경이 되야 합니다` `C는 D라는 문제가 있습니다. ` `아직 F가 결정되지 않아 E의 상태로 만들었습니다`

- **추상화정도**

적절하게 추상화하여 "누구에게나", "쉽게 이해할 수 있도록" 전달하자!

e0b8be8
`하지만, 단순한 처리는 웹 보안이 취약하며, 비밀번호 또한 암호화하지 않아 노출 될 위험이 있습니다`
**코멘트 달아 주신 것 중에 심플함은 이 메세지를 염두하신건가요?**

GitCommitMessage사이트의 [best practice](https://wiki.openstack.org/wiki/GitCommitMessages#Example_1_4)예제인데 상당히 구체적인 설명과 필요에 따라 일부 코드내용을 넣고 있네요. 물론 특수한 경우이지만 친절한 설명에 대해서 충분히 고민해볼만도 한 것 같습니다.

ef1afc4
`따라서, 비밀번호에 대한 암호화에 대해 확인할 수 있도록 추가하였습니다`
=> `따라서, 비밀번호를 암호화하여 테스트하도록 수정하였습니다`
**저 왜이렇게 `~에 대해`를 좋아하죠?**

e0b8be8
`인증이 필요한 요청에 대해서 Interceptor로 요청에 대해 식별하하도록 되어있었습니다`
=> `인증이 필요한 요청에 대해서 Interceptor로 처리하도록 되어있었습니다`
**저 왜이렇게 `~에 대해`를 좋아하죠?**

sunggyupaik commented:
[좋은 git 커밋 메시지를 작성하기 위한 8가지 약속](https://djkeh.github.io/articles/How-to-write-a-git-commit-message-kor/) 에서 아래를 우수 커밋 사례로 소개하고 있습니다.

> commit eb0b56b19017ab5c16c745e6da39c53126924ed6
>Author: Pieter Wuille <pieter.wuille@gmail.com>
>Date:   Fri Aug 1 22:57:55 2014 +0200
>
>   Simplify serialize.h's exception handling
>
>   Remove the 'state' and 'exceptmask' from serialize.h's stream
>   implementations, as well as related methods.
>
>   ........
>
>   fail(), clear(n) and exceptions() are just never called. Delete
>   them.

우리가 작성하는 변경 전, 변경 과정, 변경 결과사이에 **한줄 공백**을 넣어도 좋을 것 같습니다.

<hr>

또한,  참고해야하는 문서들이나 링크가 있을경우 다음과 같이 **참고 내용**을 추가할 수 있을 것 같습니다.

![image](https://user-images.githubusercontent.com/45138206/110920207-751f0b80-8360-11eb-9d16-95f9203d774e.png)
```java
참고
https://github.com/echoJava/book/blob/master/Java/Effective%20Java%2C%20Third%20Edition.pdf
Joshua Bloch, Effective Java (3rd),2018, Item 1. 생성자 대신 정적 팩터리 메서드를 고려하라 
```

sunggyupaik commented:
커밋 단위에 대해서는 어제의 커밋 기준으로 고민했던 점들을 적어봅니다.

3cde952
![image](https://user-images.githubusercontent.com/45138206/110927402-ecf13400-8368-11eb-8bb8-b29da02d2e9c.png)
실수로 섞여 들어간 파일을 발견하고 delete 이후 작성하였습니다. 
이건 delete 하나만 누르면 해결되고 넘어갈 일이었는데 굳이 작성을 하는게 맞는가? 생각을 했습니다.

https://github.com/CodeSoom/spring-week5-assignment-1/pull/10#discussion_r584115862
![image](https://user-images.githubusercontent.com/45138206/110930674-bd442b00-836c-11eb-9320-9bcf0b7712b8.png)
5주차 과제 피드백이였고 같은 맥락입니다. 조사를 1개 빠뜨리고 메세지는 몇줄이나 써야하다니요!!

<hr>

598c4e1
![image](https://user-images.githubusercontent.com/45138206/110927497-0befc600-8369-11eb-8f17-e7c435e60f05.png)
이 경우는 제가 주석 수정을 한번에 2개를 해버려서 묶어서 작성해버렸습니다. 굳이 나눈다면 다음과 같이 나뉩니다.
```java
토큰 해석 시, 서명이 잘못된 경우 예외가 발생한다고 되어있었습니다.
하지만 결국 발생하는 예외는 유효하지 않은 토큰에 대한 예외이므로, 예외에
따라 내용을 적는 것이 일관성이 있다고 판단하였습니다. 따라서, 예외
설명문을 수정합니다. 

또한 반환 값은 이메일이라고 되어 있었습니다.
하지만, 이는 너무 구체적인 정보라고 생각되어집니다. 따라서 사용자
정보라른 표현으로 수정하였습니다.
```
너무 간단해서 무심코 2개의 주석을 수정해버린 탓입니다.
( 하나씩 커밋 메세지를 작성하려고 다시 작업을 돌려야 할까도 고민했지만 ... ) 

하나는 토큰예외 처리 내용이고 다른 하나는 return문자의 내용인데, 크게보면 주석처리라는 공통점이 있지만, 엄밀히 말해서 독립적인 사연이 있다고 생각합니다. 하지만 , 잦은 주석의 수정을 모두 하나씩 커밋 메세지로 작성하는 것도 너무 번거롭지 않나 생각합니다. 그냥 주석을 다 수정하고 "싹 수정했음"이라고 표시할 수 있을까요 

<hr>

bf34ad9

![image](https://user-images.githubusercontent.com/45138206/110928735-8bca6000-836a-11eb-9b9e-88c17aa77e22.png)
결국 나중에는 퉁쳐서 통일성 있게 수정 했다고 한번에 작성했습니다.

sunggyupaik commented:
어떻게 하면 메세지를 좀 더 잘 써볼까 고민하다가 이런것도 찾아보네요 ㅋㅋㅋ

[후기자의 글 잘쓰는 방법 강의](https://hoos.tistory.com/188)
> 일제강점기에 들어온 일본말법 (~에 있어서, ~에 대하여, ~되어 지다)도 되도록 지양해야 합니다. 일본어 말법의 예를 들어줄게요. "현대에 있어 정보의 중요함은 말할 필요도 없다" => "현대사회에서 정보의 중요함은 말할 필요도 없다" / "후기자 의원은 그 의혹에 대해 더 이상 언급하지 말아달라고 당부했다." => "후기자 의원은 그 의혹을 더 이상 언급하지 말아달라고 당부했다."

위에서 제가 계속 반복한다고 언급했던 `~에 대해`는 일본말법이었네요!!

[정혁준의 비지니스 글쓰기](http://www.hani.co.kr/arti/economy/economy_general/927900.html) ( 한겨례뉴스 )
=> 안에 링크를 통해 총 18가지의 경우를 확인할 수 있습니다.

[일상생활 속 나도 모르게 남용하는 '번역 투'](https://www.mk.co.kr/news/society/view/2016/04/291554/) ( 매일경제 )
=> 영어 조사의 of , from, ~ing, for, by 등의 번역 투에 대한 잘못된 예시와 올바른 예시가 있습니다.

물론, 지금 우리가 사용하는 문체가 큰 문제가 있는것도 아니고, 이렇게 바꿔야하는 것도 아니지만, 참고하면 좋을 것 같습니다
(그런데, 어차피 영어로 코딩을 하기 때문에, 번역 투로 설명문이나 주석을 작성하는게 더 직관적으로 이해가 될 수도?)

잠깐 본 내용으로 한번 위에 작성한 메세지를 수정해보겠습니다. 다시 보니까 맞춤법도 엉망진창인게 많네요 퇴고의 필요성!
![image](https://user-images.githubusercontent.com/45138206/110928735-8bca6000-836a-11eb-9b9e-88c17aa77e22.png)

```java
주석 양식을 수정하라
이전에, 사용자와 상품의 생성, 수정, 삭제는 주석 양식이 각기 달랐습니다.
그래서, 비슷한 행동을 했지만 각각 다른 주석이므로 가독성이 떨어졌습니다.
따라서, 이해가 쉽고 일관성을 가지도록 수정했습니다.
```

johngrib commented:
> > Dto라고 생각하지 말고 필요한 개념을 전달한다고 생각하고 클래스를 정의해 작업해 보세요.
> 
> 이 부분에 대해서는 무조건 Dto로 보내야돼! 라는 의미보다 최대한 Entity나 주요 정보를 그대로 보내지 말고 request, response로 변환해서 보내자! 라는 생각으로 했었습니다. 단, 꼭 정해직 규칙이 있는 것이 아니기에, 너무 집착하지 않도록 하겠습니다.

각 도메인에서 의미 있는 단위를 interface로 정의해서 사용해보시는 방법도 있습니다. 저는 이런 종류의 값 제공 메소드만 갖는 인터페이스는 구현 클래스를 두지 않고 그냥 interface 안에서 static 팩토리 메소드를 두고 쓰는 방법도 종종 씁니다.

johngrib commented:
> 저는 이번 과제에서 주로 `...수정하라`를 많이 사용했는데, `고쳐라` `제거하라` `리팩토링하라` `교정하라` `변경하라` 등등 응용하면 좀 더 구분이 가능할 것 같습니다.

제 경우엔 `수정하라`를 별로 선호하지 않는데요, 뭔가를 수정했다는 것 외에 다른 의도를 알기 어렵기 때문입니다. 한편 `교정하라`는 선호하는 편입니다. `교정하라`는 뭔가 문제가 있어서 고쳤다는 느낌을 주죠. `갱신하라`같은 것도 비슷합니다.

johngrib commented:
> [e0b8be8](https://github.com/CodeSoom/spring-week7-assignment-1/commit/e0b8be892608605e37887bdb83cb015897f5dae7)
> `하지만, 단순한 처리는 웹 보안이 취약하며, 비밀번호 또한 암호화하지 않아 노출 될 위험이 있습니다`
> **코멘트 달아 주신 것 중에 심플함은 이 메세지를 염두하신건가요?**

```
인증이 필요한 요청에 대해서 Interceptor로 요청에 대해 식별하하도록
되어있었습니다. 하지만, 단순한 처리는 웹 보안이 취약하며, 비밀번호 또한
암호화하지 않아 노출 될 위험이 있습니다. 따라서, 스프링 보안을 이용하여
비밀번호를 암호화하며 인증을 하도록 수정하였습니다.
```

이렇게 작성했다면 좀 더 심플했을 것 같아요.

```
이제부터 패스워드를 플레인 텍스트로 처리하지 않고 암호화하여 처리합니다.
이 작업을 위해 스프링 시큐리티를 사용하였습니다.
```

johngrib commented:
> [e0b8be8](https://github.com/CodeSoom/spring-week7-assignment-1/commit/e0b8be892608605e37887bdb83cb015897f5dae7)
> `인증이 필요한 요청에 대해서 Interceptor로 요청에 대해 식별하하도록 되어있었습니다`
> => `인증이 필요한 요청에 대해서 Interceptor로 처리하도록 되어있었습니다`
> **저 왜이렇게 `~에 대해`를 좋아하죠?**

이 정도는 괜찮다고 생각해요. 다만 본인이 찝찝하면 더 심플하게 작성하기 위해 더 노력하시겠죠. 와 멋져라!

johngrib commented:
> 우리가 작성하는 변경 전, 변경 과정, 변경 결과사이에 **한줄 공백**을 넣어도 좋을 것 같습니다.

네 저도 그렇게 작성하시길 권장해드리고 싶어요.

johngrib commented:
> 또한, 참고해야하는 문서들이나 링크가 있을경우 다음과 같이 **참고 내용**을 추가할 수 있을 것 같습니다.
> 
> ![image](https://user-images.githubusercontent.com/45138206/110920207-751f0b80-8360-11eb-9d16-95f9203d774e.png)
> 
> ```java
> 참고
> https://github.com/echoJava/book/blob/master/Java/Effective%20Java%2C%20Third%20Edition.pdf
> Joshua Bloch, Effective Java (3rd),2018, Item 1. 생성자 대신 정적 팩터리 메서드를 고려하라 
> ```

링크를 달 때에는 다음과 같이 달아주시면 됩니다.
```
[Joshua Bloch, Effective Java (3rd),2018, Item 1. 생성자 대신 정적 팩터리 메서드를 고려하라]( https://github.com/echoJava/book/blob/master/Java/Effective%20Java%2C%20Third%20Edition.pdf )
```

그러면 이렇게 나옵니다.

[Joshua Bloch, Effective Java (3rd),2018, Item 1. 생성자 대신 정적 팩터리 메서드를 고려하라]( https://github.com/echoJava/book/blob/master/Java/Effective%20Java%2C%20Third%20Edition.pdf )

이렇게 하셔도 됩니다. 이 방법은 하나의 url을 여러 곳에서 중복적으로 사용할 때 특히 편합니다.

```
[Joshua Bloch, Effective Java (3rd),2018, Item 1. 생성자 대신 정적 팩터리 메서드를 고려하라][effective-java]

[effective-java]: https://github.com/echoJava/book/blob/master/Java/Effective%20Java%2C%20Third%20Edition.pdf
```

그러면 이렇게 나옵니다. 위랑 똑같죠?

[Joshua Bloch, Effective Java (3rd),2018, Item 1. 생성자 대신 정적 팩터리 메서드를 고려하라][effective-java]

[effective-java]: https://github.com/echoJava/book/blob/master/Java/Effective%20Java%2C%20Third%20Edition.pdf

johngrib commented:
> 그냥 주석을 다 수정하고 "싹 수정했음"이라고 표시할 수 있을까요

우리는 관공서 프로젝트를 하고 있는 게 아닙니다 :-)
자유롭게, 유연하게 생각하세요. 함께 일하는 사람이나 나를 대체하는 사람이 더 쉽게 이해할 수 있게 쓰려면 어떻게 해야 할지만 고민하세요.
때로는 과감하게 설명을 생략하기도 하고, 때로는 섬세하게 작성할 수도 있어야 합니다.

sunggyupaik committed:
사용자 수정 요청에 토큰 검사를 추가하라

이전에, 사용자 수정 요청에 아무런 토큰검사가 없었습니다.
그렇다면, 본인이 아닌 다른 사람의 토큰으로 수정 접근이 가능합니다.
따라서, 수정 요청에 검사를 하도록 권한을 추가했습니다.

sunggyupaik commented:
> 각 도메인에서 의미 있는 단위를 interface로 정의해서 사용해보시는 방법도 있습니다. 저는 이런 종류의 값 제공 메소드만 갖는 인터페이스는 구현 클래스를 두지 않고 그냥 interface 안에서 static 팩토리 메소드를 두고 쓰는 방법도 종종 씁니다.

저번에 종립님이 슬랙에 interface를 언급하셨는데 해당 부분과 연관이 있을까요?

<hr> 

>이제부터 패스워드를 플레인 텍스트로 처리하지 않고 암호화하여 처리합니다.
이 작업을 위해 스프링 시큐리티를 사용하였습니다.

굉장히 심플하고 핵심이 담겨있네요. 개인적 바램으로는 why가 조금이라도 담기면 어떨까 생각합니다. 
물론 개발자라면, 암호화한다는 의미가 보안을 강화한다고 당연히 이해할 것이고 상식적입니다.
(`암호화 => 보안 강화` 라는 명제가 맞는 표현인지는 모르겠지만요..!)
그럼에도 저는 커밋 메세지가 친절한 것이 원칙이라고 생각하여 그 이유를 보충하고 싶습니다.
```
이제부터 보안을 강화하기 위해 패스워드를 플레인 텍스트로 처리하지 않고 암호화하여 처리합니다.
이 작업을 위해 스프링 시큐리티를 사용하였습니다.
```

저는 지금껏 3문장으로 쪼개서 작성하는 것을 연습했으니 이렇게 수정해보겠습니다.
```
기존에 패스워드를 플레인 텍스트로 처리했습니다.
이제부터 보안을 강화하기 위해 암호화하여 처리합니다.
이 작업을 위해 스프링 시큐리티를 사용하였습니다.
```

누군가 두번째 문장에 구현 내용 없이 온전히 첫번째 문장의 문제점을 드러내는 것이 더 이해가 쉽다면 이렇게 해보겠습니다.
```
기존에 패스워드를 플레인 테스트로 처리했습니다.
하지만 이는 보안에 취약하며 노출의 위험이 있습니다.
따라서 스프링 시큐리티를 사용하여 암호화하였습니다.
```

누군가는 "당연한 걸 왜 그렇게 당연하게 작성했어? 내가 그것도 모를까봐?!?"라고 물어볼 수 있겠지만, 아직 연습하는 단계이므로 마치 입구에 들어 선 순간부터 모든 것을 하나하나 서빙하는 패밀리 레스토랑 종업원의 자세로 과대한 친절을 베풀었습니다.

<hr>

> 이렇게 하셔도 됩니다. 이 방법은 하나의 url을 여러 곳에서 중복적으로 사용할 때 특히 편합니다. 

감사합니다, 커밋 메세지에도 그렇게 링크를 달 수 있군요!

<hr>

> 자유롭게, 유연하게 생각하세요. 함께 일하는 사람이나 나를 대체하는 사람이 더 쉽게 이해할 수 있게 쓰려면 어떻게 해야 할지만 고민하세요. 때로는 과감하게 설명을 생략하기도 하고, 때로는 섬세하게 작성할 수도 있어야 합니다.

이것이 핵심인 것 같습니다. 과연 내가 아무것도 설명하지 않아도 주석과 메세지를 보고서 다른 사람이 이해할 수 있는가?

johngrib commented:
> 저번에 종립님이 슬랙에 interface를 언급하셨는데 해당 부분과 연관이 있을까요?

저는 class 만큼 인터페이스를 자주 쓰는 편이라 연관이 없다고도 하기도 애매하고, 있다고 하기도 애매하네요. 그냥 필요한 상황이면 사용해요.

예를 들어 다음과 같은 인터페이스를 하나 만들었다고 합시다. 이메일과 패스워드의 getter를 동시에 필요로하는 로직이 여러 군데에 있다면 이 인터페이스는 꽤 여러 곳에 달아줄 수 있습니다.

```java
public interface UserIdentifier {
  String getEmail();
  String getPassword();
}
```

그런데 굳이 구현체를 만들기 귀찮다면 저는 그냥 이렇게 정적 팩토리를 인터페이스에 넣어주기도 합니다.
```java
public interface UserIdentifier {
  String getEmail();
  String getPassword();

  static UserIdentifier of(String email, String password) {
    return new UserIdentifier() {
      @Override
      public String getEmail() {
        return email;
      }

      @Override
      public String getPassword() {
        return password;
      }
    };
  }
}
```

활용은 자신의 상상력과 경험에 달려 있습니다.

johngrib commented:
> 누군가는 "당연한 걸 왜 그렇게 당연하게 작성했어? 내가 그것도 모를까봐?!?"라고 물어볼 수 있겠지만, 아직 연습하는 단계이므로 마치 입구에 들어 선 순간부터 모든 것을 하나하나 서빙하는 패밀리 레스토랑 종업원의 자세로 과대한 친절을 베풀었습니다.

멋진 자세입니다. 화이팅!

sunggyupaik committed:
사용자 수정에서  본인 인증을 교정하라

기존에 사용자 인증은 플레인 텍스트로 하드코딩하여 검사했습니다.

하지만 수정을 시도하는 사용자에 따라 인증 정보가 달라져야 합니다.

따라서 인증토큰을 통해 자신의 이메일로 본인 인증을 하도록
교정하였습니다. 또한 이에 대한 테스트도 작성하였습니다.

sunggyupaik committed:
사용자 삭제에 로그인과 계정 검사를 추가하라

기존에 사용자 삭제는 어느 누구나 할 수 있었습니다.

하지만 안전한 삭제를 위해서 적절한 보안이 필요합니다.

따라서 로그인과 admin 계정 권한 검사를 추가하였습니다.
또한 테스트에도 검사를 추가하였습니다.

단, admin 검사는 하드코딩이므로 추후 교정이 필요합니다.

johngrib reviewed (commented):
순식간에 일요일이 되었습니다! 회고를 작성해 공유해 주세요.

sunggyupaik committed:
사용자 삭제에 admin 인증을 추가하라

기존에 사용자 삭제는 누가 요청하든지 성공했습니다.

하지만 삭제의 권한은 다른사람이 아닌 특별 권한만이 처리하도록
해야합니다.

따라서 특별 권한인 admin 계정 인증을 추가하여 삭제에 대한 조건을
만들었습니다.
